jfix-stdlib-concurrency/src/main/kotlin/ru/fix/stdlib/concurrency/events/EventReducer.kt

```diff
@@ -0,0 +1,76 @@
+package ru.fix.stdlib.concurrency.events
+
+import mu.KotlinLogging
+import java.util.concurrent.*
+
+
+/**
+ * EventReducer invokes given [handler] when [handle] function was invoked.
+ * If [handle] function was invoked (one or many times) during [handler] working,
+ * [handler] will be invoked after completing and so on.
+ *
+ * Therefore if [handle] function was invoked 100 times with breaks in 1 millisecond,
+ * and [handler]'s work takes 70 milliseconds,
+ * then [handler] wasn't invoked 100 times, but 3 times, consistently.
+ * */
+class EventReducer(
+        private val handler: () -> Unit,
+        private val shutdownCheckPeriodMs: Long = 1_000,
+        private val awaitTerminationPeriodMs: Long = 60_000
+) : AutoCloseable {
+    private val eventReceivingExecutor = Executors.newSingleThreadExecutor()
+
+    private val awaitingEventQueue = ArrayBlockingQueue<Any>(1)
+
+    fun handle() = synchronized(awaitingEventQueue) {
+        if (awaitingEventQueue.isEmpty()) {
+            awaitingEventQueue.put(Any())
+        }
+    }
+
+    fun start() {
+        CompletableFuture.runAsync(Runnable {
+            while (true) {
+                when (awaitEventOrShutdown()) {
+                    AwaitingResult.EVENT -> handler.invoke()
+                    AwaitingResult.SHUTDOWN -> return@Runnable
+                    AwaitingResult.ERROR -> {
+                    }
+                }
+            }
+        }, eventReceivingExecutor)
+    }
+
+    private enum class AwaitingResult {
+        EVENT, SHUTDOWN, ERROR
+    }
+
+    private fun awaitEventOrShutdown(): AwaitingResult {
+        try {
+            while (awaitingEventQueue.poll(shutdownCheckPeriodMs, TimeUnit.MILLISECONDS) == null) {
+                if (eventReceivingExecutor.isShutdown) {
+                    return AwaitingResult.SHUTDOWN
+                }
+            }
+        } catch (e: Exception) {
+            log.error("waiting event was interrupted", e)
+            return AwaitingResult.ERROR
+        }
+        synchronized(awaitingEventQueue) {
+            awaitingEventQueue.clear()
+        }
+        return AwaitingResult.EVENT
+    }
```

Comment:
could you please clarify why we need syncrhonized for ArrayBlockingQueue?

Reply:
my mistake was here, i removed that block of code